build error: use of undeclared identifier 'String'

Hi guys, 

I want to bind "cv::dnn::Net::getLayerNames" for Net. But I got an error:

```
opencv4nodejs git:(master) ✗ npm run build

> opencv4nodejs@4.10.1 build /Users/jansu/Development/contribution/opencv4nodejs
> node-gyp configure build

  CXX(target) Release/obj.target/opencv4nodejs/cc/modules/dnn/Net.o
In file included from ../cc/modules/dnn/Net.cc:9:
../cc/modules/dnn/NetBindings.h:97:35: error: use of undeclared identifier 'String'
      return ObjectArrayConverter<String, cv::String>::wrap(returnValue);
                                  ^
In file included from ../cc/modules/dnn/Net.cc:1:
In file included from /usr/local/include/opencv2/core.hpp:52:
In file included from /usr/local/include/opencv2/core/cvdef.h:474:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/array:109:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/algorithm:643:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/memory:3656:5: warning: destructor called on non-final 'NetBindings::SetInputWorker' that has virtual functions but non-virtual destructor [-Wdelete-non-virtual-dtor]
    __data_.second().~_Tp();
    ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/memory:3617:9: note: in instantiation of member function 'std::__1::__shared_ptr_emplace<NetBindings::SetInputWorker, std::__1::allocator<NetBindings::SetInputWorker> >::__on_zero_shared'
      requested here
        __shared_ptr_emplace(_Alloc __a, _Args&& ...__args)
        ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/memory:4277:26: note: in instantiation of function template specialization 'std::__1::__shared_ptr_emplace<NetBindings::SetInputWorker, std::__1::allocator<NetBindings::SetInputWorker>
      >::__shared_ptr_emplace<cv::dnn::experimental_dnn_34_v7::Net>' requested here
    ::new(__hold2.get()) _CntrlBlk(__a2, _VSTD::forward<_Args>(__args)...);
                         ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/memory:4656:29: note: in instantiation of function template specialization 'std::__1::shared_ptr<NetBindings::SetInputWorker>::make_shared<cv::dnn::experimental_dnn_34_v7::Net>' requested
      here
    return shared_ptr<_Tp>::make_shared(_VSTD::forward<_Args>(__args)...);
                            ^
../cc/modules/dnn/Net.cc:37:10: note: in instantiation of function template specialization 'std::__1::make_shared<NetBindings::SetInputWorker, cv::dnn::experimental_dnn_34_v7::Net>' requested here
    std::make_shared<NetBindings::SetInputWorker>(Net::Converter::unwrap(info.This())),
         ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/memory:3656:23: note: qualify call to silence this warning
    __data_.second().~_Tp();
                      ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/memory:3656:5: warning: destructor called on non-final 'NetBindings::ForwardWorker' that has virtual functions but non-virtual destructor [-Wdelete-non-virtual-dtor]
    __data_.second().~_Tp();
    ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/memory:3617:9: note: in instantiation of member function 'std::__1::__shared_ptr_emplace<NetBindings::ForwardWorker, std::__1::allocator<NetBindings::ForwardWorker> >::__on_zero_shared'
      requested here
        __shared_ptr_emplace(_Alloc __a, _Args&& ...__args)
        ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/memory:4277:26: note: in instantiation of function template specialization 'std::__1::__shared_ptr_emplace<NetBindings::ForwardWorker, std::__1::allocator<NetBindings::ForwardWorker>
      >::__shared_ptr_emplace<cv::dnn::experimental_dnn_34_v7::Net>' requested here
    ::new(__hold2.get()) _CntrlBlk(__a2, _VSTD::forward<_Args>(__args)...);
                         ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/memory:4656:29: note: in instantiation of function template specialization 'std::__1::shared_ptr<NetBindings::ForwardWorker>::make_shared<cv::dnn::experimental_dnn_34_v7::Net>' requested here
    return shared_ptr<_Tp>::make_shared(_VSTD::forward<_Args>(__args)...);
                            ^
../cc/modules/dnn/Net.cc:53:10: note: in instantiation of function template specialization 'std::__1::make_shared<NetBindings::ForwardWorker, cv::dnn::experimental_dnn_34_v7::Net>' requested here
    std::make_shared<NetBindings::ForwardWorker>(Net::Converter::unwrap(info.This())),
         ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/memory:3656:23: note: qualify call to silence this warning
    __data_.second().~_Tp();
                      ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/memory:3656:5: warning: destructor called on non-final 'NetBindings::GetLayerNamesWorker' that has virtual functions but non-virtual destructor [-Wdelete-non-virtual-dtor]
    __data_.second().~_Tp();
    ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/memory:3617:9: note: in instantiation of member function 'std::__1::__shared_ptr_emplace<NetBindings::GetLayerNamesWorker, std::__1::allocator<NetBindings::GetLayerNamesWorker>
      >::__on_zero_shared' requested here
        __shared_ptr_emplace(_Alloc __a, _Args&& ...__args)
        ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/memory:4277:26: note: in instantiation of function template specialization 'std::__1::__shared_ptr_emplace<NetBindings::GetLayerNamesWorker,
      std::__1::allocator<NetBindings::GetLayerNamesWorker> >::__shared_ptr_emplace<cv::dnn::experimental_dnn_34_v7::Net>' requested here
    ::new(__hold2.get()) _CntrlBlk(__a2, _VSTD::forward<_Args>(__args)...);
                         ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/memory:4656:29: note: in instantiation of function template specialization 'std::__1::shared_ptr<NetBindings::GetLayerNamesWorker>::make_shared<cv::dnn::experimental_dnn_34_v7::Net>'
      requested here
    return shared_ptr<_Tp>::make_shared(_VSTD::forward<_Args>(__args)...);
                            ^
../cc/modules/dnn/Net.cc:69:12: note: in instantiation of function template specialization 'std::__1::make_shared<NetBindings::GetLayerNamesWorker, cv::dnn::experimental_dnn_34_v7::Net>' requested here
      std::make_shared<NetBindings::GetLayerNamesWorker>(Net::Converter::unwrap(info.This())),
           ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/memory:3656:23: note: qualify call to silence this warning
    __data_.second().~_Tp();
                      ^
3 warnings and 1 error generated.
make: *** [Release/obj.target/opencv4nodejs/cc/modules/dnn/Net.o] Error 1
gyp ERR! build error
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/Users/jansu/.nvm/versions/node/v8.11.2/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:262:23)
gyp ERR! stack     at emitTwo (events.js:126:13)
gyp ERR! stack     at ChildProcess.emit (events.js:214:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:198:12)
gyp ERR! System Darwin 18.2.0
gyp ERR! command "/Users/jansu/.nvm/versions/node/v8.11.2/bin/node" "/Users/jansu/.nvm/versions/node/v8.11.2/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "configure" "build"
gyp ERR! cwd /Users/jansu/Development/contribution/opencv4nodejs
gyp ERR! node -v v8.11.2
gyp ERR! node-gyp -v v3.8.0
gyp ERR! not ok
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! opencv4nodejs@4.10.1 build: `node-gyp configure build`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the opencv4nodejs@4.10.1 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/jansu/.npm/_logs/2018-12-04T10_32_18_004Z-debug.log
```

Here code snippet in my NetBindings.h:
```
  struct GetLayerNamesWorker : public CatchCvExceptionWorker {
  public:
    cv::dnn::Net self;
    GetLayerNamesWorker(cv::dnn::Net self) {
      this->self = self;
    }

    std::vector<cv::String> returnValue;

    std::string executeCatchCvExceptionWorker() {
      returnValue = self.getLayerNames();
      return "";
    }

    v8::Local<v8::Value> getReturnValue(){
      return ObjectArrayConverter<String, cv::String>::wrap(returnValue);
    }
  };
```
Does anynoe know what's going on? 

Any help would be appreciated.

After I tried  "StringArrayConverter::wrap()" the build got the other error.

Here is full log for build:

```
➜  opencv4nodejs git:(master) ✗ npm run build

> opencv4nodejs@4.11.0 build /Users/jansu/Development/contribution/opencv4nodejs
> node-gyp configure build

  CXX(target) Release/obj.target/opencv4nodejs/cc/modules/dnn/Net.o
In file included from ../cc/modules/dnn/Net.cc:9:
../cc/modules/dnn/NetBindings.h:97:41: error: no viable conversion from 'vector<cv::String, allocator<cv::String>>' to
      'vector<std::__1::basic_string<char>, allocator<std::__1::basic_string<char>>>'
      return StringArrayConverter::wrap(layerNames);
                                        ^~~~~~~~~~
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/vector:534:5: note: candidate constructor not viable: no
      known conversion from 'std::vector<cv::String>' to 'const std::__1::vector<std::__1::basic_string<char>,
      std::__1::allocator<std::__1::basic_string<char> > > &' for 1st argument
    vector(const vector& __x);
    ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/vector:541:5: note: candidate constructor not viable: no
      known conversion from 'std::vector<cv::String>' to 'initializer_list<std::__1::vector<std::__1::basic_string<char>,
      std::__1::allocator<std::__1::basic_string<char> > >::value_type>' (aka 'initializer_list<std::__1::basic_string<char> >') for 1st argument
    vector(initializer_list<value_type> __il);
    ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/vector:547:5: note: candidate constructor not viable: no
      known conversion from 'std::vector<cv::String>' to 'std::__1::vector<std::__1::basic_string<char>, std::__1::allocator<std::__1::basic_string<char> > >
      &&' for 1st argument
    vector(vector&& __x)
    ^
/Users/jansu/Development/contribution/opencv4nodejs/node_modules/native-node-utils/src/AbstractConverter.h:59:37: note: passing argument to parameter 'val'
      here
        static v8::Local<v8::Value> wrap(T val) {
                                           ^
In file included from ../cc/modules/dnn/Net.cc:1:
In file included from /usr/local/include/opencv2/core.hpp:52:
In file included from /usr/local/include/opencv2/core/cvdef.h:474:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/array:109:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/algorithm:643:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/memory:3656:5: warning: destructor called on non-final
      'NetBindings::SetInputWorker' that has virtual functions but non-virtual destructor [-Wdelete-non-virtual-dtor]
    __data_.second().~_Tp();
    ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/memory:3617:9: note: in instantiation of member function
      'std::__1::__shared_ptr_emplace<NetBindings::SetInputWorker, std::__1::allocator<NetBindings::SetInputWorker> >::__on_zero_shared' requested here
        __shared_ptr_emplace(_Alloc __a, _Args&& ...__args)
        ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/memory:4277:26: note: in instantiation of function template
      specialization 'std::__1::__shared_ptr_emplace<NetBindings::SetInputWorker, std::__1::allocator<NetBindings::SetInputWorker>
      >::__shared_ptr_emplace<cv::dnn::experimental_dnn_34_v7::Net>' requested here
    ::new(__hold2.get()) _CntrlBlk(__a2, _VSTD::forward<_Args>(__args)...);
                         ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/memory:4656:29: note: in instantiation of function template
      specialization 'std::__1::shared_ptr<NetBindings::SetInputWorker>::make_shared<cv::dnn::experimental_dnn_34_v7::Net>' requested here
    return shared_ptr<_Tp>::make_shared(_VSTD::forward<_Args>(__args)...);
                            ^
../cc/modules/dnn/Net.cc:38:10: note: in instantiation of function template specialization 'std::__1::make_shared<NetBindings::SetInputWorker,
      cv::dnn::experimental_dnn_34_v7::Net>' requested here
    std::make_shared<NetBindings::SetInputWorker>(Net::Converter::unwrap(info.This())),
         ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/memory:3656:23: note: qualify call to silence this warning
    __data_.second().~_Tp();
                      ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/memory:3656:5: warning: destructor called on non-final
      'NetBindings::ForwardWorker' that has virtual functions but non-virtual destructor [-Wdelete-non-virtual-dtor]
    __data_.second().~_Tp();
    ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/memory:3617:9: note: in instantiation of member function
      'std::__1::__shared_ptr_emplace<NetBindings::ForwardWorker, std::__1::allocator<NetBindings::ForwardWorker> >::__on_zero_shared' requested here
        __shared_ptr_emplace(_Alloc __a, _Args&& ...__args)
        ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/memory:4277:26: note: in instantiation of function template
      specialization 'std::__1::__shared_ptr_emplace<NetBindings::ForwardWorker, std::__1::allocator<NetBindings::ForwardWorker>
      >::__shared_ptr_emplace<cv::dnn::experimental_dnn_34_v7::Net>' requested here
    ::new(__hold2.get()) _CntrlBlk(__a2, _VSTD::forward<_Args>(__args)...);
                         ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/memory:4656:29: note: in instantiation of function template
      specialization 'std::__1::shared_ptr<NetBindings::ForwardWorker>::make_shared<cv::dnn::experimental_dnn_34_v7::Net>' requested here
    return shared_ptr<_Tp>::make_shared(_VSTD::forward<_Args>(__args)...);
                            ^
../cc/modules/dnn/Net.cc:54:10: note: in instantiation of function template specialization 'std::__1::make_shared<NetBindings::ForwardWorker,
      cv::dnn::experimental_dnn_34_v7::Net>' requested here
    std::make_shared<NetBindings::ForwardWorker>(Net::Converter::unwrap(info.This())),
         ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/memory:3656:23: note: qualify call to silence this warning
    __data_.second().~_Tp();
                      ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/memory:3656:5: warning: destructor called on non-final
      'NetBindings::GetLayerNamesWorker' that has virtual functions but non-virtual destructor [-Wdelete-non-virtual-dtor]
    __data_.second().~_Tp();
    ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/memory:3617:9: note: in instantiation of member function
      'std::__1::__shared_ptr_emplace<NetBindings::GetLayerNamesWorker, std::__1::allocator<NetBindings::GetLayerNamesWorker> >::__on_zero_shared' requested
      here
        __shared_ptr_emplace(_Alloc __a, _Args&& ...__args)
        ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/memory:4277:26: note: in instantiation of function template
      specialization 'std::__1::__shared_ptr_emplace<NetBindings::GetLayerNamesWorker, std::__1::allocator<NetBindings::GetLayerNamesWorker>
      >::__shared_ptr_emplace<cv::dnn::experimental_dnn_34_v7::Net>' requested here
    ::new(__hold2.get()) _CntrlBlk(__a2, _VSTD::forward<_Args>(__args)...);
                         ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/memory:4656:29: note: in instantiation of function template
      specialization 'std::__1::shared_ptr<NetBindings::GetLayerNamesWorker>::make_shared<cv::dnn::experimental_dnn_34_v7::Net>' requested here
    return shared_ptr<_Tp>::make_shared(_VSTD::forward<_Args>(__args)...);
                            ^
../cc/modules/dnn/Net.cc:70:12: note: in instantiation of function template specialization 'std::__1::make_shared<NetBindings::GetLayerNamesWorker,
      cv::dnn::experimental_dnn_34_v7::Net>' requested here
      std::make_shared<NetBindings::GetLayerNamesWorker>(Net::Converter::unwrap(info.This())),
           ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/memory:3656:23: note: qualify call to silence this warning
    __data_.second().~_Tp();
                      ^
3 warnings and 1 error generated.
make: *** [Release/obj.target/opencv4nodejs/cc/modules/dnn/Net.o] Error 1
gyp ERR! build error
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/Users/jansu/.nvm/versions/node/v8.11.2/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:262:23)
gyp ERR! stack     at emitTwo (events.js:126:13)
gyp ERR! stack     at ChildProcess.emit (events.js:214:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:198:12)
gyp ERR! System Darwin 18.2.0
gyp ERR! command "/Users/jansu/.nvm/versions/node/v8.11.2/bin/node" "/Users/jansu/.nvm/versions/node/v8.11.2/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "configure" "build"
gyp ERR! cwd /Users/jansu/Development/contribution/opencv4nodejs
gyp ERR! node -v v8.11.2
gyp ERR! node-gyp -v v3.8.0
gyp ERR! not ok
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! opencv4nodejs@4.11.0 build: `node-gyp configure build`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the opencv4nodejs@4.11.0 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/jansu/.npm/_logs/2018-12-06T22_51_01_184Z-debug.log
```
Here is my code snapshot with line number:
![screen shot 2018-12-07 at 6 39 31 am](https://user-images.githubusercontent.com/1734406/49616518-18e41900-f9ec-11e8-91bc-1b873480beae.png)

--- a/cc/modules/dnn/Net.cc
+++ b/cc/modules/dnn/Net.cc
@@ -20,6 +20,10 @@ NAN_MODULE_INIT(Net::Init) {
   Nan::SetPrototypeMethod(ctor, "setInputAsync", SetInputAsync);
   Nan::SetPrototypeMethod(ctor, "forward", Forward);
   Nan::SetPrototypeMethod(ctor, "forwardAsync", ForwardAsync);
+  Nan::SetPrototypeMethod(ctor, "getLayerNames", GetLayerNames);
+  Nan::SetPrototypeMethod(ctor, "getLayerNamesAsync", GetLayerNamesAsync);
+  Nan::SetPrototypeMethod(ctor, "getUnconnectedOutLayers", GetUnconnectedOutLayers);
+  Nan::SetPrototypeMethod(ctor, "getUnconnectedOutLayersAsync", GetUnconnectedOutLayersAsync);
 
   target->Set(Nan::New("Net").ToLocalChecked(), ctor->GetFunction());
 };
@@ -61,6 +65,34 @@ NAN_METHOD(Net::ForwardAsync) {
     "Net::ForwardAsync",
     info
   );
+}
+
+NAN_METHOD(Net::GetLayerNames) {
+  FF::SyncBinding(
+      std::make_shared<NetBindings::GetLayerNamesWorker>(Net::Converter::unwrap(info.This())),
+      "Net::GetLayerNames",
+      info);
+}
+
+NAN_METHOD(Net::GetLayerNamesAsync) {
+  FF::AsyncBinding(
+      std::make_shared<NetBindings::GetLayerNamesWorker>(Net::Converter::unwrap(info.This())),
+      "Net::GetLayerNamesAsync",
+      info);
+}
+
+NAN_METHOD(Net::GetUnconnectedOutLayers) {
+  FF::SyncBinding(
+      std::make_shared<NetBindings::GetUnconnectedOutLayersWorker>(Net::Converter::unwrap(info.This())),
+      "Net::GetUnconnectedOutLayers",
+      info);
+}
+
+NAN_METHOD(Net::GetUnconnectedOutLayersAsync) {
+  FF::AsyncBinding(
+      std::make_shared<NetBindings::GetUnconnectedOutLayersWorker>(Net::Converter::unwrap(info.This())),
+      "Net::GetUnconnectedOutLayersAsync",
+      info);
 }
 
 #endif

--- a/cc/modules/dnn/Net.h
+++ b/cc/modules/dnn/Net.h
@@ -28,6 +28,10 @@ public:
 	static NAN_METHOD(SetInputAsync);
 	static NAN_METHOD(Forward);
 	static NAN_METHOD(ForwardAsync);
+  static NAN_METHOD(GetLayerNames);
+  static NAN_METHOD(GetLayerNamesAsync);
+  static NAN_METHOD(GetUnconnectedOutLayers);
+  static NAN_METHOD(GetUnconnectedOutLayersAsync);
 };
 
 #endif

--- a/cc/modules/dnn/NetBindings.h
+++ b/cc/modules/dnn/NetBindings.h
@@ -79,6 +79,47 @@ namespace NetBindings {
     }
   };
 
+  struct GetLayerNamesWorker : public CatchCvExceptionWorker {
+  public:
+    cv::dnn::Net self;
+    GetLayerNamesWorker(cv::dnn::Net self) {
+      this->self = self;
+    }
+
+    std::vector<std::string> returnValue;
+
+    std::string executeCatchCvExceptionWorker() {
+      std::vector<cv::String> layerNames = self.getLayerNames();
+      std::vector<std::string> strings(
+        layerNames.begin(),
+        layerNames.end());
+      returnValue = strings;
+      return "";
+    }
+
+    v8::Local<v8::Value> getReturnValue() {
+      return StringArrayConverter::wrap(returnValue);
+    }
+  };
+
+  struct GetUnconnectedOutLayersWorker : public CatchCvExceptionWorker {
+  public:
+    cv::dnn::Net self;
+    GetUnconnectedOutLayersWorker(cv::dnn::Net self) {
+      this->self = self;
+    }
+
+    std::vector<int> layerIndexes;
+
+    std::string executeCatchCvExceptionWorker() {
+      layerIndexes = self.getUnconnectedOutLayers();
+      return "";
+    }
+
+    v8::Local<v8::Value> getReturnValue() {
+      return IntArrayConverter::wrap(layerIndexes);
+    }
+  };
 
 }
 

--- a/cc/modules/dnn/dnn.cc
+++ b/cc/modules/dnn/dnn.cc
@@ -20,8 +20,8 @@ NAN_MODULE_INIT(Dnn::Init) {
   Nan::SetMethod(target, "blobFromImages", BlobFromImages);
   Nan::SetMethod(target, "blobFromImagesAsync", BlobFromImagesAsync);
 #if CV_VERSION_MINOR > 3
-  Nan::SetMethod(target, "readNetFromDarknet", readNetFromDarknet);
-  Nan::SetMethod(target, "readNetFromDarknetAsync", readNetFromDarknetAsync);
+  Nan::SetMethod(target, "readNetFromDarknet", ReadNetFromDarknet);
+  Nan::SetMethod(target, "readNetFromDarknetAsync", ReadNetFromDarknetAsync);
   Nan::SetMethod(target, "NMSBoxes", NMSBoxes);
 #endif
 };
@@ -92,19 +92,17 @@ NAN_METHOD(Dnn::BlobFromImagesAsync) {
 }
 
 #if CV_VERSION_MINOR > 3
-NAN_METHOD(Dnn::readNetFromDarknet)
-{
+NAN_METHOD(Dnn::ReadNetFromDarknet) {
   FF::SyncBinding(
-      std::make_shared<DnnBindings::readNetFromDarknetWorker>(),
-      "readNetFromDarknet",
+      std::make_shared<DnnBindings::ReadNetFromDarknetWorker>(),
+      "ReadNetFromDarknet",
       info);
 }
 
-NAN_METHOD(Dnn::readNetFromDarknetAsync)
-{
+NAN_METHOD(Dnn::ReadNetFromDarknetAsync) {
   FF::AsyncBinding(
-      std::make_shared<DnnBindings::readNetFromDarknetWorker>(),
-      "readNetFromDarknetAsync",
+      std::make_shared<DnnBindings::ReadNetFromDarknetWorker>(),
+      "ReadNetFromDarknetAsync",
       info);
 }
 

--- a/cc/modules/dnn/dnn.h
+++ b/cc/modules/dnn/dnn.h
@@ -20,8 +20,8 @@ public:
   static NAN_METHOD(BlobFromImages);
   static NAN_METHOD(BlobFromImagesAsync);
 #if CV_VERSION_MINOR > 3
-  static NAN_METHOD(readNetFromDarknet);
-  static NAN_METHOD(readNetFromDarknetAsync);
+  static NAN_METHOD(ReadNetFromDarknet);
+  static NAN_METHOD(ReadNetFromDarknetAsync);
   static NAN_METHOD(NMSBoxes);
 #endif
 };

--- a/cc/modules/dnn/dnnBindings.h
+++ b/cc/modules/dnn/dnnBindings.h
@@ -6,7 +6,7 @@
 namespace DnnBindings {
 
 #if CV_VERSION_MINOR > 3
-  struct readNetFromDarknetWorker : public CatchCvExceptionWorker{
+  struct ReadNetFromDarknetWorker : public CatchCvExceptionWorker{
   public:
     std::string cfgFile;
     std::string darknetModelFile = "";

--- a/examples/dnnDarknetYOLORealTimeObjectDetection.js
+++ b/examples/dnnDarknetYOLORealTimeObjectDetection.js
@@ -4,7 +4,7 @@
  */
 const fs = require("fs");
 const path = require("path");
-const { cv, drawBlueRect, runVideoDetection } = require("./utils");
+const { cv, runVideoDetection } = require("./utils");
 
 if (!cv.xmodules.dnn) {
   throw new Error("exiting: opencv4nodejs compiled without dnn module");
@@ -41,6 +41,13 @@ const labels = fs
 
 // initialize tensorflow darknet model from modelFile
 const net = cv.readNetFromDarknet(cfgFile, weightsFile);
+const allLayerNames = net.getLayerNames();
+const unconnectedOutLayers = net.getUnconnectedOutLayers();
+
+// determine only the *output* layer names that we need from YOLO
+const layerNames = unconnectedOutLayers.map(layerIndex => {
+  return allLayerNames[layerIndex - 1];
+});
 
 const classifyImg = img => {
   // object detection model works with 416 x 416 images
@@ -49,11 +56,8 @@ const classifyImg = img => {
   const [imgHeight, imgWidth] = img.sizes;
 
   // network accepts blobs as input
-  const inputBlob = cv.blobFromImage(img, 1 / 255.0, size, vec3, true, true);
+  const inputBlob = cv.blobFromImage(img, 1 / 255.0, size, vec3, true, false);
   net.setInput(inputBlob);
-
-  // specify two layers "yolo_16" and "yolo_23"
-  const layerNames = ["yolo_16", "yolo_23"];
 
   console.time("net.forward");
   // forward pass input through entire network
@@ -95,20 +99,25 @@ const classifyImg = img => {
 
         indices.forEach(i => {
           const rect = boxes[i];
-          const imgRect = new cv.Rect(rect.x, rect.y, rect.width, rect.height);
-          drawBlueRect(img, imgRect);
+
+          const pt1 = new cv.Point(rect.x, rect.y);
+          const pt2 = new cv.Point(rect.x + rect.width, rect.y + rect.height);
+          const rectColor = new cv.Vec(255, 0, 0);
+          const rectThickness = 2;
+          const rectLineType = cv.LINE_8;
+
+          // draw the rect for the object
+          img.drawRectangle(pt1, pt2, rectColor, rectThickness, rectLineType);
+
           const text = labels[classIDs[i]];
-          img.putText(
-            text,
-            new cv.Point(rect.x, rect.y + 0.1 * imgHeight),
-            cv.FONT_ITALIC,
-            2,
-            {
-              color: new cv.Vec(255, 0, 0),
-              thickness: 2
-            }
-          );
-          drawBlueRect(img, imgRect);
+          const org = new cv.Point(rect.x, rect.y + 15);
+          const fontFace = cv.FONT_HERSHEY_SIMPLEX;
+          const fontScale = 0.5;
+          const textColor = new cv.Vec(123, 123, 255);
+          const thickness = 2;
+
+          // put text on the object
+          img.putText(text, org, fontFace, fontScale, textColor, thickness);
         });
       }
     });

--- a/examples/dnnTensorflowObjectDetection.js
+++ b/examples/dnnTensorflowObjectDetection.js
@@ -5,7 +5,7 @@
 const fs = require("fs");
 const path = require("path");
 const classNames = require("./dnnTensorflowObjectDetectionClassNames");
-const { cv, drawBlueRect, runVideoDetection } = require("./utils");
+const { cv, runVideoDetection } = require("./utils");
 
 if (!cv.xmodules.dnn) {
   throw new Error("exiting: opencv4nodejs compiled without dnn module");
@@ -51,7 +51,7 @@ const classifyImg = img => {
 
   // get height and width from the image
   const [imgHeight, imgWidth] = img.sizes;
-  const numRows = outputBlob.sizes.slice(2,3);
+  const numRows = outputBlob.sizes.slice(2, 3);
 
   for (let y = 0; y < numRows; y += 1) {
     const confidence = outputBlob.at([0, 0, y, 2]);
@@ -62,22 +62,25 @@ const classifyImg = img => {
       const boxY = imgHeight * outputBlob.at([0, 0, y, 4]);
       const boxWidht = imgWidth * outputBlob.at([0, 0, y, 5]);
       const boxHeight = imgHeight * outputBlob.at([0, 0, y, 6]);
-      const imgRect = new cv.Rect(boxX, boxY, boxWidht, boxHeight);
 
-      // draw the blue rect for the object
-      drawBlueRect(img, imgRect);
+      const pt1 = new cv.Point(boxX, boxY);
+      const pt2 = new cv.Point(boxWidht, boxHeight);
+      const rectColor = new cv.Vec(23, 230, 210);
+      const rectThickness = 2;
+      const rectLineType = cv.LINE_8;
+
+      // draw the rect for the object
+      img.drawRectangle(pt1, pt2, rectColor, rectThickness, rectLineType);
+
+      const text = `${className} ${confidence.toFixed(5)}`;
+      const org = new cv.Point(boxX, boxY + 15);
+      const fontFace = cv.FONT_HERSHEY_SIMPLEX;
+      const fontScale = 0.5;
+      const textColor = new cv.Vec(255, 0, 0);
+      const thickness = 2;
 
       // put text on the object
-      img.putText(
-        className,
-        new cv.Point(boxX, boxY + 0.1 * imgHeight),
-        cv.FONT_ITALIC,
-        2,
-        {
-          color: new cv.Vec(255, 0, 0),
-          thickness: 2
-        }
-      );
+      img.putText(text, org, fontFace, fontScale, textColor, thickness);
     }
   }
 

